### PR TITLE
Formproc

### PR DESCRIPTION
**Pull request by semseysandor**
**Extension core functionality**
- create placeholder groups, tags
- import form-processor
- update group, tag IDs to real ones

**Next steps**
- add unit tests
- add permission
- add functionality: remove previous user opt-out

### CODE REVIEW

**Comment by semseysandor** [comment]
Note:
CI will fail until https://github.com/reflexive-communications/rc-base/pull/17 get merged

File: processor/newsletter_subscription.json [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                "phone_is_primary": "0",
+                "phone_location_type": "1",
+                "phone_update_existing": "0"
+            },
+            "mapping": {
+                "contact_id": "action.save_email.contact_id",
+                "first_name": "input.first_name",
+                "last_name": "input.last_name"
+            },
+            "condition_configuration": "",
```
Comment (akosgarai) [inline processor/newsletter_subscription.json:97]: The first name and last name are optional parameters according to the input configuration. They could be empty. If the contact with the email exists then the first name and last name could be overwritten with empty string. I think the not empty condition could be used to eliminate this issue.

Comment (akosgarai) [reply processor/newsletter_subscription.json]: You don't need to do separate actions. The `CheckParameters` configuration could be used here. See this example: https://github.com/reflexive-communications/contribution-processor/blob/main/templates/processor-template.json#L140-L160 You can setup multiple parameters and `one of is not empty` function here to tell at least one parameter has to be set.

File: CRM/FormprocNewsletter/Upgrader.php [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    /**
+     * Create/update dummy group
+     *
+     * @throws \API_Exception
+     * @throws \CRM_Core_Exception
+     * @throws \Civi\API\Exception\UnauthorizedException
+     */
+    protected function installPlaceholderGroup(array $group): void
+    {
+        $group_id = CRM_RcBase_Api_Get::groupIDByName($group['name']);
```
Comment (akosgarai) [inline CRM/FormprocNewsletter/Upgrader.php:75]: I'm not sure that the default group should be connected to a given name. As far as i see you can't update the name on the UI forms, but you can set the title.

File: processor/newsletter_subscription.json (hunk 1 above)
Comment (semseysandor) [reply processor/newsletter_subscription.json]: That's cool, I haven't noticed that option

File: CRM/FormprocNewsletter/Upgrader.php (hunk 2 above)
Comment (semseysandor) [reply CRM/FormprocNewsletter/Upgrader.php]: I switched it to group name

**Comment by semseysandor** [comment]
- Added group, tag description
- Changed group name to title (looks for specified title not name)
- moved Form Processor spec to `/templates`
- removed unit-tests from CI workflow

Now it should be ready to merge

**Review by akosgarai — APPROVED** [review]
Just one thing to mention about this process. With this setup, the CRM admins are not allowed to update the title of the placeholder groups, or the name of the placeholder tags. If they do, then a disable/enable step will create the tags and groups again, and due to this, the form processor config will be also updated. I think the option for updating the names for something less meaningless will be a valid request. The contribution-processor [saves the group and tag ids to the setting](https://github.com/reflexive-communications/contribution-processor/blob/main/CRM/Contributionprocessor/Upgrader.php#L32-L35). On enable it checks that the stuff with the given id exists or not. It creates a new one only if the id [does not exist](https://github.com/reflexive-communications/contribution-processor/blob/main/CRM/Contributionprocessor/Upgrader.php#L61-L64).

**Comment by semseysandor** [comment]
My concept was, that these placeholder groups and tags are there just to be able to import the Form-processor. After import (installation, or enabling, if it was deleted) admins should update the form-processor to the actual newsletters (there are 2 in the template, but you can add more, or delete), tags and groups. These tags and groups are not handled by the extension as they're client specific.

It's basically an install and config workflow. Placeholders are like the defaults, which should be changed.
This means it won't work out-of-the-box, but it wasn't the goal either.

**Comment by semseysandor** [comment]
I will merge this now, later I will update to save placeholder group and tag IDs to DB.

**Pull request merged**